ITISFoundation/osparc-simcore
# 🏗️ Upgrades repository to python 3.8.10

## What do these changes do?

- Upgrades all python code in repository, i.e. all packages, tests and services [except ``director`` whose development is frozen], to **python 3.8.10**. See [what is new](https://docs.python.org/3/whatsnew/3.8.html)
- Repository's version in ``requirements/PYTHON_VERSION`` ( repo checked using ``tests/environment-setup/test_used_python.py``)
- CI runs in 3.8.10 except for ``python-linting`` that is run in both **3.8** and **3.9**


### Main changes

- *Dockerfile* base images
- ``python_requires`` specs on ``setup.py`` and other metadata/checks
- *requirement* listings were re-compile using latest [pip-tools](https://github.com/jazzband/pip-tools) 
   - compilation run within controlled environment ``requirements/tools/Dockerfile``
   - NO package was upgraded
   - Notice how all backport libraries (e.g. dataclasses) were automatically removed  (SEE ``requirements/packages-notes.md`` for more details)
-  config files  in tooling (e.g. ``.pre-commit-config.yaml``) etc
- Upgrade python in CI setup
- web-server's ``garbage_collector.py`` ignores cancellation and enters in an errors loop. I am still not sure what is the cause so I introduced a mechanism to enforce getting out of the infinite ``while True`` loop.

### Some highlights in 3.8

- [assignment expressions](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions) @mguidon, I know you were waiting for this!
- new features in [asyncio](https://docs.python.org/3/whatsnew/3.8.html#asyncio) @sanderegg @GitHK  interesting:
    -  ``asyncio.run(coro())`` 
    -  ``python -m asyncio`` creates an asyncio REPL 3.8.0  
    - The exception asyncio.CancelledError now inherits from BaseException rather than Exception
    - tasks can now be named (see garbage-collector background task). Handy to avoid duplicates or to search ...
- time benchmark [from 3.5 to 3.8](https://docs.python.org/3/whatsnew/3.8.html#demos-and-tools)

## Related issue/s

- #1412 updates to py 3.8
- #1388 


## How to test

- RECOMMENDED: install [pyenv](https://github.com/pyenv/pyenv) in your system
- Check available versions: ``pyenv install --list | grep 3.8``
- install 3.8: ``pyenv install 3.8.6``
- see all installed versions: ``pyenv versions``
- use 3.8 in ``osparc-simcore`` project. Will create a ``.python-version`` file: 
```cmd
cd osparc-simcore
pyenv local 3.8.10
python --version
```
- create devenv
```cmd
 make devenv
 source .venv/bin/activate
```

## Checklist
- [ ] when this is upgraded review #1884
- [x] check speed difference between 3.8 and 3.6? and major changes
- [x] CI matrix
- [x] Upgrades python service's Dockerfile base
- [x] Adapt code for [new features](https://docs.python.org/3/whatsnew/3.8.html)
   - ``CancelledError`` is not in ``asyncio.future`` anymore
   - New [AsyncMock](https://docs.python.org/3/whatsnew/3.8.html#unittest). No need to use the trick of returning a ``Future`` to emulate a coroutine. 
   - ``pylint`` (has some issues with 3.9)
